### ISSUE TITLE
[Question] Custom environment: First episode in evaluation always perform poorly

### ISSUE THREAD

**Issue by RaffaeleGalliera** [DLR-RM/rl-baselines3-zoo#289]
When running evaluations during training or enjoy.py the first episode always performs poorly, while the subsequent episodes are more in line with the expected performance. Initially I thought it was just a random behavior, but I can notice it every time now. Is there any known problem that might cause this?

**System Info**
Describe the characteristic of your environment:
 * Stable Baselines3 is installed via pip on a Docker container
 * GPU used: 1x 3080 mobile
 * Python 3.8.13
 * PyTorch version 1.12.0

**Additional context**
I am using SAC with TimeLimit, History wrappers and negative rewards. The environment is a bit complicated to explain as there are some external processes that are spawned at the beginning of each episode and cleaned up when reset is called.

**Comment by JakobThumm** [comment]
I have a similar problem. For me, the evaluation runs for >100 episodes, even though I set the number of eval episodes to 3.

I was able to reproduce the error for a common environment:
```
python train.py --algo sac --env BipedalWalkerHardcore-v3 --yaml-file hyperparameters/sac.yml -P --seed 42 --eval-freq 5000 --eval-episodes 3 --n-eval-envs 1
```
`sac.yml`
```
BipedalWalkerHardcore-v3:
  env_wrapper:
    - gym.wrappers.TimeLimit:
        max_episode_steps: 1000
  n_timesteps: !!float 1e7
  policy: 'MlpPolicy'
  learning_rate: lin_7.3e-4
  buffer_size: 1000000
  batch_size: 256
  ent_coef: 0.005
  gamma: 0.99
  tau: 0.01
  train_freq: 1
  gradient_steps: 1
  learning_starts: 10000
  policy_kwargs: "dict(net_arch=[256, 256])"
```
Note that this issue occurs if and only if I change the `net_arch` from `[400, 300]` to `[256, 256]`. This issue also does not occur on seed 0, but it does happen on seed 42.

Apparently, the evaluation is doing more than I expect. I would assume, the evaluation just runs for the given number of episodes and then continues training.

**Comment by RaffaeleGalliera** [comment]
I did some further checks, retraining the agent different times also changing rewards and hyperparameters. 
The issue actually appears only when using enjoy.py and not when evaluating the during training. So my bad, when running evaluations during training everything looks fine.
Doing some further investigations I have also noticed that the agent actually acts way differently from how it was doing throughout all the training process, actions are different compared to training and achieve significant lower rewards (on avg. it achieved ~-120 during training, but when using enjoy.py we are lucky to get something above -150 and in average is way lower -150). 

The environment is normalized and is set as an infinite-horizon task (I do use TimeLimit) and I save checkpoints every 100K steps and tried all of them out of curiosity, but every policy performs  poorly (including best). It is worth to mention that the environment has a certain degree of stochasticity, but I exclude the problem as after 1M steps (5000 episodes) the mean reward is consistent.
Could this be somehow linked to normalization statistics? Any idea what I could be missing or if there is anything I can try?

Also: The task I use for evaluation is different from the one used during training, and all this considerations are done directly using the training task, so I can exclude that poor performance was linked to the eval/testing task being different. Please let me know if any additional detail could be helpful! Thanks!

Below the mean reward graph during training:
![image](https://user-images.githubusercontent.com/23361949/196061301-541daca3-417e-4043-b801-147b8dc79f5b.png)

**Comment by araffin** [comment]
> Below the mean reward graph during training:

probably related to https://github.com/DLR-RM/stable-baselines3/issues/1063, you should try with stochastic controller at eval time `--stochastic` with the enjoy script.

> The environment is normalized 

you mean you are using `VecNormalize`?

**Comment by chengwei-xia** [comment]
> Could you please show some details of the evaluation (enjoy.py)? I am facing a very similar issue as yours.
